pow/src/main/java/tech/pegasys/artemis/pow/Eth1Manager.java

```diff
@@ -13,113 +13,121 @@
-                // find first valid block in history
-                return findFirstValidBlockInHistory(block);
-              } else if (comparison < 0) {
-                // If block timestamp is less than min genesis time
-                // subscribe to new block events and wait for the first
-                // valid block
-                return waitForFirstValidBlock();
-              } else {
-                return SafeFuture.completedFuture(block);
-              }
-            })
-        .thenAccept(this::publishFirstValidBlock)
-        .exceptionallyCompose(
-            err -> {
-              if (latestBlockDisposable != null) {
-                latestBlockDisposable.dispose();
-              }
+        .thenCompose(eth1Provider::getEth1BlockFuture);
+  }
 
-              LOG.debug(
-                  "Eth1MinGenesisTimeBlockFinder failed to find first valid block. Retry in "
-                      + Constants.ETH1_MIN_GENESIS_TIME_BLOCK_RETRY_TIMEOUT
-                      + " seconds",
-                  err);
-
-              System.out.println("adding one delayed run");
-              return asyncRunner.runAfterDelay(
-                  this::findAndPublishFirstValidBlock,
-                  Constants.ETH1_MIN_GENESIS_TIME_BLOCK_RETRY_TIMEOUT,
-                  TimeUnit.SECONDS);
-            });
+  public SafeFuture<Boolean> isHeadBeforeMinGenesis(SafeFuture<EthBlock.Block> headBlock) {
+    return headBlock.thenApply(
+        block -> {
+          int comparison = compareBlockTimestampToMinGenesisTime(block);
+          // If block timestamp is greater than min genesis time,
+          // min genesis block must have been in history
+          return comparison > 0;
+        });
   }
 
-  private void publishFirstValidBlock(EthBlock.Block block) {
-    minGenesisTimeBlockEventChannel.onMinGenesisTimeBlock(
+  public static EthBlock.Block postMinGenesisTimeBlock(
+      Eth1EventsChannel eth1EventsChannel, EthBlock.Block block) {
+    eth1EventsChannel.onMinGenesisTimeBlock(
         new MinGenesisTimeBlockEvent(
             UnsignedLong.valueOf(block.getTimestamp()),
             UnsignedLong.valueOf(block.getNumber()),
             Bytes32.fromHexString(block.getHash())));
+    return block;
   }
 
   /**
-   * Find first valid block in history that has timestamp greater than MIN_GENESIS_TIME
+   * Find first block in history that has timestamp greater than MIN_GENESIS_TIME
    *
    * @param estimationBlock estimationBlock that will be used for estimation
-   * @return first valid block in history
+   * @return min genesis time block
    */
-  private SafeFuture<EthBlock.Block> findFirstValidBlockInHistory(EthBlock.Block estimationBlock) {
-    UnsignedLong estimatedFirstValidBlockNumber =
-        getEstimatedFirstValidBlockNumber(estimationBlock, Constants.SECONDS_PER_ETH1_BLOCK);
+  private SafeFuture<EthBlock.Block> findMinGenesisTimeBlockInHistory(
```

Comment:
I'd be tempted to split this into a `MinimumGenesisTimeBlockFinder` class.

Reply:
done.